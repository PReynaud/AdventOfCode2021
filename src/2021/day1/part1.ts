import { isIncreasing } from './common';

export function countNumberOfIncrease(values: number[]): number {
  let result = 0;

  values.forEach((element, index) => {
    if (index > 0) {
      if (isIncreasing(values[index - 1], values[index])) {
        result++;
      }
    }
  });

  return result;
}
