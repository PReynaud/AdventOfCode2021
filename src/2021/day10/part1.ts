export function simplifyLine(value: string): string {
  let currentValue = value;
  let previousValue = currentValue;
  do {
    previousValue = currentValue;
    currentValue = currentValue
      .replace('()', '')
      .replace('[]', '')
      .replace('{}', '')
      .replace('<>', '');
  } while (currentValue && currentValue.length < previousValue.length);
  return currentValue;
}

const openingChars = ['[', '{', '(', '<'];

export function isIncomplete(line: string): boolean {
  return line.split('').every((char) => openingChars.includes(char));
}

function findFirstWrongChar(line: string): string {
  for (let i = 0; i < line.length; i++) {
    if (!openingChars.includes(line[i])) {
      return line[i];
    }
  }
  return '';
}

export function simplifyAndFind(line: string): string {
  return findFirstWrongChar(simplifyLine(line));
}

export function selectIncorrectLines(allLines: string[]): string[] {
  return allLines.filter((oneLine) => !isIncomplete(simplifyLine(oneLine)));
}

function returnValueForChar(char: string): number {
  if (char === ')') return 3;
  if (char === ']') return 57;
  if (char === '}') return 1197;
  if (char === '>') return 25137;
  return 0;
}

export function countIncorrectLines(allLines: string[]): number {
  return selectIncorrectLines(allLines)
    .map(simplifyAndFind)
    .map(returnValueForChar)
    .reduce((acc, value) => acc + value, 0);
}
