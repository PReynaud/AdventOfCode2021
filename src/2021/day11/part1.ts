interface Case {
  value: number;
  hasFlashed: boolean;
}

export type Grid = Case[][];

export function transformRawValuesToGrid(rawValues: string[]): Grid {
  return rawValues.map((oneRow) =>
    oneRow.split('').map((oneValue) => ({
      value: Number(oneValue),
      hasFlashed: false,
    }))
  );
}

interface Position {
  x: number;
  y: number;
}

export function findNeighbours(
  currentPosition: Position,
  xMax: number,
  yMax: number
): Position[] {
  const x = currentPosition.x;
  const y = currentPosition.y;

  return [
    { x: x - 1, y: y - 1 },
    { x: x - 1, y },
    { x: x - 1, y: y + 1 },
    { x, y: y - 1 },
    { x, y: y + 1 },
    { x: x + 1, y: y - 1 },
    { x: x + 1, y },
    { x: x + 1, y: y + 1 },
  ].filter(
    (onePosition) =>
      onePosition.x >= 0 &&
      onePosition.y >= 0 &&
      onePosition.x < xMax &&
      onePosition.y < yMax
  );
}

function getCaseForPosition(grid: Grid, position: Position): Case {
  return grid[position.y][position.x];
}

function increaseValue(grid: Grid, currentPosition: Position): void {
  const currentCase = getCaseForPosition(grid, currentPosition);
  if (currentCase.value < 9 && !currentCase.hasFlashed) {
    currentCase.value++;
  } else if (currentCase.value === 9 && !currentCase.hasFlashed) {
    currentCase.value = 0;
    currentCase.hasFlashed = true;
    increaseNeighbours(grid, currentPosition);
  }
}

export function increaseNeighbours(
  grid: Grid,
  currentPosition: Position
): void {
  const neighbours = findNeighbours(
    currentPosition,
    grid[0].length,
    grid.length
  );
  neighbours.forEach((neighbour) => {
    increaseValue(grid, neighbour);
  });
}

function resetAndCount(grid: Grid): number {
  let result = 0;
  grid
    .flat()
    .flat()
    .forEach((oneCase) => {
      if (oneCase.value === 0) {
        oneCase.hasFlashed = false;
        result++;
      }
    });

  return result;
}

function displayGrid(grid: Grid): void {
  let result = '';
  for (let y = 0; y < grid.length; y++) {
    result += grid[y].map((oneValue) => oneValue.value).join('') + '\n';
  }
  console.log(result);
}

export function runOneStep(grid: Grid): number {
  for (let x = 0; x < grid[0].length; x++) {
    for (let y = 0; y < grid.length; y++) {
      increaseValue(grid, { x, y });
    }
  }

  return resetAndCount(grid);
}

export function runSteps(rawValues: string[], numberOfSteps: number): number {
  const grid = transformRawValuesToGrid(rawValues);
  let numberOfFlashes = 0;
  for (let count = 0; count < numberOfSteps; count++) {
    numberOfFlashes += runOneStep(grid);
  }
  return numberOfFlashes;
}
