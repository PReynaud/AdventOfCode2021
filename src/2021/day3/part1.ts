import { convertBinaryToNumber } from './common';

export function extractGammaRate(values: string[]): string {
  if (values[0]) {
    const maxIndex = values[0].length;
    let result = '';
    for (let i = 0; i < maxIndex; i++) {
      let zeroValues = 0;
      let oneValues = 0;

      values.forEach((oneString) => {
        if (oneString[i] === '0') {
          zeroValues++;
        } else if (oneString[i] === '1') {
          oneValues++;
        }
      });

      result += zeroValues > oneValues ? '0' : '1';
    }
    return result;
  }
  return '';
}

export function invertBites(binaryString: string): string {
  let result = '';
  for (let i = 0; i < binaryString.length; i++) {
    result += binaryString[i] === '0' ? '1' : '0';
  }
  return result;
}

export function extractRates(values: string[]): [string, string] {
  const gammaRate = extractGammaRate(values);
  const epsilonRate = invertBites(gammaRate);
  return [gammaRate, epsilonRate];
}

export function calculatePowerConsumption(values: string[]): number {
  const [gammaRate, epsilonRate] = extractRates(values);
  return convertBinaryToNumber(gammaRate) * convertBinaryToNumber(epsilonRate);
}
