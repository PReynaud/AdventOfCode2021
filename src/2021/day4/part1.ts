import {
  BoardCase,
  calculateScoreWinningBoard,
  checkBoardIsWinner,
  extractBoards,
  extractDrawnNumbers,
  markValuesFromDrawnNumber,
} from './common';

function run(boards: BoardCase[][][], initialNumbers: number[]): number | null {
  let hasWinning = false;
  let numbersToDraw = initialNumbers.slice();
  let currentDrawnNumber: number | null = null;

  while (!hasWinning && numbersToDraw.length > 0) {
    currentDrawnNumber = numbersToDraw[0];
    boards.forEach((board) =>
      markValuesFromDrawnNumber(board, currentDrawnNumber!)
    );
    hasWinning = boards.some((board) => checkBoardIsWinner(board));
    numbersToDraw = numbersToDraw.slice(1);
  }

  if (hasWinning) {
    const winningBoard = boards.find((board) => checkBoardIsWinner(board));
    const score = calculateScoreWinningBoard(winningBoard!);
    return score * currentDrawnNumber!;
  }

  return null;
}

export function extractAndRun(rawValues: string[]): number | null {
  const drawnNumbers = extractDrawnNumbers(rawValues);
  const boards = extractBoards(rawValues);

  return run(boards, drawnNumbers);
}
