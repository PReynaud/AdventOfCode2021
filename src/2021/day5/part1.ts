import {
  calculateOverlapping,
  createEmptyGrid,
  drawStraightLine,
  extractLineInformations,
  Grid,
  isStraightLine,
  Line,
} from './common';

export function drawAllLines(grid: Grid, lines: Line[]): void {
  lines
    .filter((oneLine) => isStraightLine(oneLine))
    .forEach((oneLine) => {
      drawStraightLine(grid, oneLine);
    });
}

export function calculateFromRawValues(rawValues: string[]): number {
  const extracted = extractLineInformations(rawValues);
  const grid = createEmptyGrid(extracted);
  drawAllLines(grid, extracted);
  return calculateOverlapping(grid);
}
