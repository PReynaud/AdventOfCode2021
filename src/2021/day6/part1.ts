import { extractValuesFromRaw } from './common';

export function decreaseValuesAndAddNewFishes(values: number[]): number[] {
  let numberOfNewFishes = 0;
  const newValues = values.map((oneValue) => {
    if (oneValue === 0) {
      numberOfNewFishes++;
      return 6;
    }
    return oneValue - 1;
  });
  return [...newValues, ...Array(numberOfNewFishes).fill(8)];
}

export function calculateNumberOfFish(
  rawValues: string,
  numberOfDays: number
): number {
  let values = extractValuesFromRaw(rawValues);
  for (let i = 0; i < numberOfDays; i++) {
    values = decreaseValuesAndAddNewFishes(values);
  }
  return values.length;
}
