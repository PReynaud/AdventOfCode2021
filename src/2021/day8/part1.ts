import { extractSecondPartValuesFromRaw } from './common';

export function countNumberOfAppearances(rawValues: string[]): number {
  const values = extractSecondPartValuesFromRaw(rawValues);

  const allowedLengthValues = [2, 3, 4, 7];
  const result = values
    .flat()
    .flat()
    .filter((value) => allowedLengthValues.includes(value.length)).length;

  return result;
}
