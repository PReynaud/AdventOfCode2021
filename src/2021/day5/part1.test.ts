import {
  calculateOverlapping,
  createEmptyGrid,
  displayGrid,
  drawStraightLine,
  extractLineInformations,
  extractOneLine,
} from './common';
import { calculateFromRawValues, drawAllLines } from './part1';

describe('Day 5 - Part 1', () => {
  const exampleRaw = [
    '0,9 -> 5,9',
    '8,0 -> 0,8',
    '9,4 -> 3,4',
    '2,2 -> 2,1',
    '7,0 -> 7,4',
    '6,4 -> 2,0',
    '0,9 -> 2,9',
    '3,4 -> 1,4',
    '0,0 -> 8,8',
    '5,5 -> 8,2',
  ];
  it('should correctly extract one line', () => {
    expect(extractOneLine('0,9 -> 5,9')).toEqual({
      x1: 0,
      y1: 9,
      x2: 5,
      y2: 9,
    });
    expect(extractOneLine('421,642 -> 91,312')).toEqual({
      x1: 421,
      y1: 642,
      x2: 91,
      y2: 312,
    });
  });

  it('should extract all lines', () => {
    const extracted = extractLineInformations(exampleRaw);
    expect(extracted).toHaveLength(10);
    expect(extracted[0]).toEqual({
      x1: 0,
      y1: 9,
      x2: 5,
      y2: 9,
    });
  });

  it('should create empty grid', () => {
    const extracted = extractLineInformations(exampleRaw);
    const grid = createEmptyGrid(extracted);
    expect(grid).toHaveLength(10);
    grid.forEach((oneRow) => {
      expect(oneRow).toHaveLength(10);
    });
    expect(grid).toMatchSnapshot();
  });

  it('should draw a line', () => {
    const extracted = extractLineInformations(exampleRaw);
    const grid = createEmptyGrid(extracted);

    let line = { x1: 0, y1: 9, x2: 5, y2: 9 };
    drawStraightLine(grid, line);
    expect(displayGrid(grid)).toMatchSnapshot();

    line = { x1: 0, y1: 4, x2: 6, y2: 4 };
    drawStraightLine(grid, line);
    expect(displayGrid(grid)).toMatchSnapshot();

    line = { x1: 0, y1: 4, x2: 0, y2: 6 };
    drawStraightLine(grid, line);
    expect(displayGrid(grid)).toMatchSnapshot();
  });

  it('should draw all lines', () => {
    const extracted = extractLineInformations(exampleRaw);
    const grid = createEmptyGrid(extracted);
    drawAllLines(grid, extracted);

    expect(displayGrid(grid)).toMatchSnapshot();
  });

  it('should calculate overlapping points', () => {
    const extracted = extractLineInformations(exampleRaw);
    const grid = createEmptyGrid(extracted);
    drawAllLines(grid, extracted);

    expect(calculateOverlapping(grid)).toEqual(5);
  });

  it('should validate example', () => {
    const result = calculateFromRawValues(exampleRaw);
    expect(result).toEqual(5);
  });
});
