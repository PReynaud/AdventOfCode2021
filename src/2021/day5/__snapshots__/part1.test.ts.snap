// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day 5 - Part 1 should create empty grid 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`Day 5 - Part 1 should draw a line 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
1 1 1 1 1 1 0 0 0 0
"
`;

exports[`Day 5 - Part 1 should draw a line 2`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
1 1 1 1 1 1 1 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
1 1 1 1 1 1 0 0 0 0
"
`;

exports[`Day 5 - Part 1 should draw a line 3`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
2 1 1 1 1 1 1 0 0 0
1 0 0 0 0 0 0 0 0 0
1 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
1 1 1 1 1 1 0 0 0 0
"
`;

exports[`Day 5 - Part 1 should draw all lines 1`] = `
"0 0 0 0 0 0 0 1 0 0
0 0 1 0 0 0 0 1 0 0
0 0 1 0 0 0 0 1 0 0
0 0 0 0 0 0 0 1 0 0
0 1 1 2 1 1 1 2 1 1
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
2 2 2 1 1 1 0 0 0 0
"
`;
